HubSpot.NET/Core/HubSpotBaseClient.cs

```diff
@@ -1,128 +1,198 @@
-using System;
-using System.Collections.Generic;
-using Flurl;
-using HubSpot.NET.Core.Interfaces;
-using HubSpot.NET.Core.Requests;
-using RestSharp;
-
 namespace HubSpot.NET.Core
 {
+    using HubSpot.NET.Api.OAuth.Dto;
+    using HubSpot.NET.Core.Interfaces;
+    using HubSpot.NET.Core.Serializers;
+    using Newtonsoft.Json;
+    using RestSharp;
+    using System.Collections.Generic;
+
     public class HubSpotBaseClient : IHubSpotClient
     {
-        private readonly RequestSerializer _serializer = new RequestSerializer(new RequestDataConverter());
         private readonly RestClient _client;
 
         private string _baseUrl => "https://api.hubapi.com";
+        private readonly HubSpotAuthenticationMode _mode;
+        
+        public string BasePath { get => _baseUrl; }
+
+        // Used for HAPIKEY method
+        private readonly string _apiKeyName = "hapikey";
         private readonly string _apiKey;
 
-        public HubSpotBaseClient(string apiKey)
-        {
+        // Used for OAUTH
```

Comment:
There were a few reasons @rezalas and I decided to go this route. There are several reasons why we decided to take the BaseClient this direction, the main one though is the ease of client configuration for the developers coding against this SDK.

We added an overloaded constructor to the BaseClient so that they have the option to authenticate how they wish and have the same basic implementation available to them regardless of authentication style. This also helps us, as maintainers, not have separate locations to keep up with or sync up for common functionality. Furthermore, the functionality of a HAPIKey based client and an OAuth based client would not be different enough to justify having them separate; this is mostly due to the Authentication process being the only difference.

Additionally, the interface should be used to give someone else a contract to make their own implementation of the BaseClient if their requirements call for something other than what HS.NET provides. You are right though, that usually people would not need to use the two authentication styles together, but this change wasn't intended to enable them to be used together; rather it was intended to make the authentication style easier to configure at the BaseClient's constructor and then HS.NET will handle the rest internally from one location. For instance, in our current usage of HS.NET we have a factory creating a `HubSpotApi` instance which supplies that instance to all of our classes which consume it.

Does that help clear up what we were going for?

Reply:
In addition to that reasoning, the HS team appears to be slowly moving away from their hapikey implementation, and in fact all the new endpoints require oauth instead. I don't think they're going to drop support for hapikey anytime soon but I tried to be forward thinking and provide the most functionality from the base client.

Reply:
OK cool - I'm happy to pull this into a few of our projects and see how that fits in.

I wasn't aware about the overall move to OAuth - I'd always thought that was intended for use with the Timeline APIs (and others) where a HS user must grant permission over their HS portal to the application.

Happy to roll with this though! 👍